drivers/charger/charger_bq25180.c

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * BQ25180 Datasheet: https://www.ti.com/lit/gpn/bq25180
+ */
+
+#define DT_DRV_COMPAT ti_bq25180
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/charger.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(bq25180, CONFIG_CHARGER_LOG_LEVEL);
+
+#define BQ25180_STAT0 0x00
+#define BQ25180_STAT1 0x01
+#define BQ25180_FLAG0 0x02
+#define BQ25180_VBAT_CTRL 0x03
+#define BQ25180_ICHG_CTRL 0x04
+#define BQ25180_IC_CTRL 0x07
+#define BQ25180_SHIP_RST 0x09
+#define BQ25180_MASK_ID 0x0c
+
+#define BQ25180_ICHG_CHG_DIS BIT(7)
+#define BQ25180_ICHG_MSK GENMASK(6, 0)
+#define BQ25180_WATCHDOG_SEL_1_MSK GENMASK(1, 0)
+#define BQ25180_WATCHDOG_DISABLE 0x03
+#define BQ25180_DEVICE_ID_MSK GENMASK(3, 0)
+#define BQ25180_DEVICE_ID 0x00
+#define BQ25180_SHIP_RST_EN_RST_SHIP_MSK GENMASK(6, 5)
+#define BQ25180_SHIP_RST_EN_RST_SHIP_ADAPTER 0x20
+#define BQ25180_SHIP_RST_EN_RST_SHIP_BUTTON 0x40
+
+struct bq25180_config {
+	struct i2c_dt_spec i2c;
+	uint32_t initial_current_microamp;
+};
+
+/*
+ * For ICHG <= 35mA = ICHGCODE + 5mA
+ * For ICHG > 35mA = 40 + ((ICHGCODE-31)*10)mA.
+ * Maximum programmable current = 1000mA
+ *
+ * Return: value between 0 and 127, negative on error.
+ */
+static int bq25180_ma_to_ichg(uint32_t current_ma, uint8_t *ichg)
+{
+	if (current_ma < 5 || current_ma > 1000) {
+		return -EINVAL;
+	}
```

Comment:
minor nit. I'm more of a fan of using the `CLAMP` macro add issuing a log warning rather than erroring out, but I'm not opposed to this either.

Reply:
good idea, guess that makes it easier to share code between devices, done